### PR TITLE
feat(exhook): expose headers for on_messages_publish hook

### PR DESCRIPTION
**Pull request by HJianBo**
We have added more parameters for the 'OnMessagePublish' callback to support better customizability.

For the `Message`  object of `exhook.proto`, we have added the `headers` property, which includes:

```protobuff
message Message {
  ...

  // The key of header can be:
  //  - username:
  //    * Readonly
  //    * The username of sender client
  //    * Value type: utf8 string
  //  - protocol:
  //    * Readonly
  //    * The protocol name of sender client
  //    * Value type: string enum with "mqtt", "mqtt-sn", ...
  //  - peerhost:
  //    * Readonly
  //    * The peerhost of sender client
  //    * Value type: ip address string
  //  - allow_publish:
  //    * Writable
  //    * Whether to allow the message to be published by emqx
  //    * Value type: string enum with "true", "false", default is "true"
  //
  // Notes: All header may be missing, which means that the message does not
  //   carry these headers. We can guarantee that clients coming from MQTT,
  //   MQTT-SN, CoAP, LwM2M and other natively supported protocol clients will
  //   carry these headers, but there is no guarantee that messages published
  //   by other means will do, e.g. messages published by HTTP-API
  map<string, string> headers = 8;
}
```

We also modified the implementations of MQTT-SN, CoAP, LwM2M, Stomp, ExProto so that they all take these headers when publishing a message

### CODE REVIEW

File: apps/emqx_exhook/src/emqx_exhook_server.erl [hunk 1]
```diff
@@ -174,16 +181,18 @@ resovle_hookspec(HookSpecs) when is_list(HookSpecs) ->
...
-                        case lists:member(Name, MessageHooks) of
-                            true ->
-                                Acc#{Name => #{topics => maps:get(topics, HookSpec, [])}};
-                            _ ->
-                                Acc#{Name => #{}}
-                        end;
-                    _ -> error({unknown_hookpoint, Name})
+                Name = try
+                           binary_to_existing_atom(Name0, utf8)
+                       catch T:R:_ -> {T,R}
```
Comment (zmstone) [inline apps/emqx_exhook/src/emqx_exhook_server.erl:186]:
```suggestion
                       catch T:R -> {T,R}
```

File: apps/emqx_coap/src/emqx_coap.app.src [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {application, emqx_coap,
  [{description, "EMQ X CoAP Gateway"},
-  {vsn, "4.3.0"}, % strict semver, bump manually!
+  {vsn, "4.3.1"}, % strict semver, bump manually!
```
Comment (tigercl) [inline apps/emqx_coap/src/emqx_coap.app.src:3]: 4.4.0？

File: apps/emqx_lwm2m/src/emqx_lwm2m.app.src [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {application,emqx_lwm2m,
              [{description,"EMQ X LwM2M Gateway"},
-              {vsn, "4.3.4"}, % strict semver, bump manually!
+              {vsn, "4.3.5"}, % strict semver, bump manually!
```
Comment (zmstone) [inline apps/emqx_lwm2m/src/emqx_lwm2m.app.src:3]: maybe bump to 4.4.0

File: apps/emqx_stomp/src/emqx_stomp.app.src [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {application, emqx_stomp,
  [{description, "EMQ X Stomp Protocol Plugin"},
-  {vsn, "4.3.2"}, % strict semver, bump manually!
+  {vsn, "4.3.3"}, % strict semver, bump manually!
```
Comment (zmstone) [inline apps/emqx_stomp/src/emqx_stomp.app.src:3]: 4.4.0

**Review by zmstone — CHANGES_REQUESTED** [review]

File: apps/emqx_coap/src/emqx_coap.app.src (hunk 2 above)
[the remark this answers is not in the document]
Comment (HJianBo) [reply apps/emqx_coap/src/emqx_coap.app.src:3]: It's not a feature, it's more like a bugfix or enhancement. I'll sync the changes to this application to the main-4.3 branch later.

**Comment by HJianBo** [comment]
Some commits have a port to main-v4.3
https://github.com/emqx/emqx/pull/6139

**Review by terry-xiaoyu — APPROVED** [review]

**Pull request merged**